api/src/main/java/org/pmiops/workbench/api/CloudTaskUserController.java

```diff
@@ -127,67 +136,117 @@ public ResponseEntity<Void> auditProjectAccessBatch(List<Long> userIds) {
+      return ResponseEntity.noContent().build();
     }
-    log.info(String.format("Successfully synchronized %d users", userIds.size()));
-    return ResponseEntity.noContent().build();
+  }
+
+  private int syncOneUser(DbUser user) {
+    try {
+      // 2FA synchronization requires an outgoing call to gsuite. For this reason, we
+      // optimize to only verify that users who have 2FA enabled, still have it enabled.
+      // Users who don't have 2FA will go through an active flow to enable it, and are not
+      // dependent on this offline check. Disabled users have no access anyways, so don't
+      // bother checking them either.
+      if (!user.getDisabled()) {
+        Optional<AccessModuleStatus> status =
+            accessModuleService.getAccessModuleStatus(user, DbAccessModuleName.TWO_FACTOR_AUTH);
+        if (status.isPresent() && status.get().getCompletionEpochMillis() != null) {
+          user = userService.syncTwoFactorAuthStatus(user, Agent.asSystem());
+        }
+      }
+
+      // Note: each module synchronization calls updateUserAccessTiers() which checks
+      // the status of *all* modules, so this serves as a general fallback as well (e.g. due to
+      // partial system failures or bugs), ensuring that the database and access tier groups are
+      // consistent with access module statuses.
+      userService.syncDuccVersionStatus(user, Agent.asSystem());
+    } catch (WorkbenchException e) {
+      log.log(Level.SEVERE, "Failed to synchronize access for user " + user.getUsername(), e);
+      return 1; // error
+    }
+    return 0; // no error
   }
 
   /**
-   * Takes in batch of user Ids check whether users have initial credits that have expired
+   * Takes in batch of user IDs and check whether users have initial credits that have expired
    *
    * @param userIds : Batch of user IDs from cloud task queue: checkCreditsExpirationForUserIDsQueue
-   * @return
    */
   @Override
   public ResponseEntity<Void> checkCreditsExpirationForUserIDsBatch(List<Long> userIds) {
-    log.info(String.format("Checking initial credits expiration for %d users...", userIds.size()));
-    initialCreditsService.checkCreditsExpirationForUserIDs(userIds);
-    log.info("Done checking initial credits expiration.");
+    return processUserIdBatch(
+        userIds,
+        "checking initial credits expiration",
+        initialCreditsService::checkCreditsExpirationForUserIDs);
+  }
+
+  /**
+   * Takes in batch of user IDs and check whether users' access has expired
+   *
+   * @param userIds : Batch of user IDs from cloud task queue: accessExpirationEmailQueue
+   */
+  @Override
+  public ResponseEntity<Void> sendAccessExpirationEmailsBatch(List<Long> userIds) {
+    processDbUserBatch(
+        userIds,
+        "checking access expiration",
+        dbUsers -> {
+          userService.maybeSendAccessTierExpirationEmails(dbUsers);
+          return 0;
```

Comment:
a little klugey... this one doesn't count errors, but it still needs to report an error count

Reply:
Seems reasonable